File: apps/flutter-app/lib/core/error/failures.dart
import 'package:equatable/equatable.dart';

sealed class Failure extends Equatable {
  final String message;

  const Failure(this.message);

  @override
  List<Object?> get props => [message];
}

final class ServerFailure extends Failure {
  const ServerFailure([super.message = 'Ha ocurrido un error en el servidor']);
}

final class InsufficientBalanceFailure extends Failure {
  const InsufficientBalanceFailure([
    super.message = 'No tiene saldo disponible para vincularse al fondo',
  ]);
}

final class AlreadySubscribedFailure extends Failure {
  const AlreadySubscribedFailure([
    super.message = 'Ya se encuentra vinculado a este fondo',
  ]);
}

final class NotFoundFailure extends Failure {
  const NotFoundFailure([super.message = 'Recurso no encontrado']);
}

File: apps/angular-app/src/app/core/errors/failures.ts
export abstract class Failure {
  constructor(public readonly message: string) {}
}

export class ServerFailure extends Failure {
  constructor(message: string = 'Error del servidor. Intente nuevamente.') {
    super(message);
  }
}

export class NetworkFailure extends Failure {
  constructor(message: string = 'Error de conexion. Verifique su internet.') {
    super(message);
  }
}

export class InsufficientBalanceFailure extends Failure {
  constructor(fundName: string, minAmount: number) {
    super(`No tiene saldo suficiente para vincularse al fondo ${fundName}. Monto minimo: $${minAmount.toLocaleString('es-CO')}`);
  }
}

export class AlreadySubscribedFailure extends Failure {
  constructor(fundName: string) {
    super(`Ya se encuentra vinculado al fondo ${fundName}`);
  }
}

export class NotFoundFailure extends Failure {
  constructor(message: string = 'Recurso no encontrado.') {
    super(message);
  }
}

export class ValidationFailure extends Failure {
  constructor(message: string) {
    super(message);
  }
}
